### PR TITLE
[JEP-228] Unforking XStream

### PR DESCRIPTION
**Pull request by jglick**
[JEP-228](https://jenkins.io/jep/228)

### Proposed changelog entries

* Jenkins now uses an updated version of the XStream serialization library without custom patches.

### Proposed upgrade guidelines

https://www.jenkins.io/blog/2020/11/10/spring-xstream/

See the [compatibility table](https://github.com/jenkinsci/jep/blob/master/jep/228/compatibility.adoc).

### Desired reviewers

@jenkinsci/core

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [x] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by jglick** [comment]
Seems that the defense in #3248 is no longer necessary as of https://github.com/x-stream/xstream/commit/d0ad410c6c2e20b93b28cd4793da83c82b4560e8, though it is harmless to leave in.

File: core/src/main/java/hudson/tools/ToolInstallation.java
Comment (jglick) [inline core/src/main/java/hudson/tools/ToolInstallation.java:84]: This is a problem. `DescribableList` is not `Serializable` so removing `transient` will prevent this class from being serialized over Remoting. Yet keeping `transient` will prevent tool installations from saving their properties in XML. I have not found any way to get the effect of `XStreamSerializable` without patching `PureJavaReflectionProvider`.

Comment (jglick) [reply core/src/main/java/hudson/tools/ToolInstallation.java:84]: (I am not sure why `ToolInstallation` was ever made `Serializable` to begin with. The normal usage is to have it define some `PATH` entry on the controller side, which does not require the `ToolInstallation` itself to ever be sent to the agent.)

Comment (jglick) [reply core/src/main/java/hudson/tools/ToolInstallation.java:84]: Also the current annotation was explicitly rejected upstream: https://github.com/x-stream/xstream/pull/70#issuecomment-254899878

File: core/src/main/java/hudson/tools/ToolInstallation.java
Comment (daniel-beck) [inline core/src/main/java/hudson/tools/ToolInstallation.java:233]: Nice!

Comment (jglick) [reply core/src/main/java/hudson/tools/ToolInstallation.java:233]: Well, not nice, but seems to work well enough. I am trying to fix popular plugins to not need this hack.

File: core/src/main/java/hudson/util/XStream2.java
Comment (jglick) [inline core/src/main/java/hudson/util/XStream2.java:261]: May need to be adjusted for XStream 1.5.x: https://github.com/x-stream/xstream/issues/207

Comment (joehni) [reply core/src/main/java/hudson/util/XStream2.java:261]: No. This is the same for XStream 1.4.x and 1.5.x. With this line you overwrite any default initialization and you are initializing the XStream Security Framework with a blacklist without any entries i.e. you allow anything. Your BlacklistedTypesConverter is now responsibly to cover any of the CVEs.

Comment (jglick) [reply core/src/main/java/hudson/util/XStream2.java:261]: In our case that is fine since we are actually using a whitelist, despite the name, so this line merely ensures that the current behavior is maintained and that XStream does not print a warning to stderr. https://jenkins.io/jep/200 FTR

File: bom/pom.xml
Comment (jglick) [inline bom/pom.xml:329]: If merged, we should also archive both https://github.com/jenkinsci/xstream and https://github.com/jenkinsci/xstream-fork.

File: core/src/main/java/hudson/diagnosis/OldDataMonitor.java
Comment (jglick) [inline core/src/main/java/hudson/diagnosis/OldDataMonitor.java:212]: Very handy for diagnosing test failures related to XStream!

File: core/src/main/java/hudson/util/PersistedList.java
Comment (jglick) [inline core/src/main/java/hudson/util/PersistedList.java:186]: Useful for PCT, until we can get most plugins moved onto new POMs.

File: core/src/test/java/hudson/util/XStream2Test.java
Comment (jglick) [inline core/src/test/java/hudson/util/XStream2Test.java:544]: Just demonstrating what would happen with plugins that used annotations without a `processAnnotations` call.

File: pom.xml
Comment (jglick) [inline pom.xml:93]: @daniel-beck please release me!

File: test/src/test/java/hudson/model/ProjectTest.java
Comment (jglick) [inline test/src/test/java/hudson/model/ProjectTest.java:259]: Needed for JTH upgrade, unrelated to XStream.

File: test/src/test/java/hudson/util/XStream2AnnotationTest.java
Comment (jglick) [inline test/src/test/java/hudson/util/XStream2AnnotationTest.java:65]: More realistic demonstration of compatibility impact for plugins of turning off autodetection.

**Comment by jglick** [comment]
I consider this eligible for merge as soon as the next LTS baseline has been finalized, code review is concluded, and https://github.com/jenkinsci/matrix-auth-plugin/pull/89 has been released so I can include that version as detached.

**Review by timja — APPROVED** [review]
all looks fine, a number of trivial changes could be detached to a separate PR and merged quickly, but the PR isn't that large.

**Review by res0nance — APPROVED** [review]

**Comment by jglick** [comment]
Would prefer to get at least the following released first:

* https://github.com/jenkinsci/matrix-auth-plugin/pull/89 @daniel-beck 
* https://github.com/jenkinsci/email-ext-plugin/pull/220 some owner TBD (@basil?)
* https://github.com/jenkinsci/github-plugin/pull/239 some owner TBD (@lanwen? @integer? @KostyaSha despite not appearing to have deploy permission in RPU?)

**Review by oleg-nenashev — DISMISSED** [review]
Please follow the pull request template. Changelog and upgrade guide drafts are required for this pull request

**Review by jvz — APPROVED** [review]
LGTM, seems to work with random plugins I tried out. Read through the associated JEP-228 docs as well.

**Comment by timja** [comment]
Is anyone else wanting to review this or should we ship in the next weekly?

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: pom.xml
Comment (basil) [inline pom.xml:93]:
```suggestion
    <matrix-auth.version>2.6.4</matrix-auth.version>
```

**Review by basil — APPROVED** [review]

**Comment by daniel-beck** [comment]
> This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Given the PR size, might make sense to wait a few days longer, since the next weekly is still a week off anyway.

File: pom.xml
Comment (jglick) [reply pom.xml:93]: Oh right, forgot about that! Thanks.

**Comment by jvz** [comment]
I'll be merging this after #4848 is merged.

**Comment by jglick** [comment]
Cleanup: [INFRA-2794](https://issues.jenkins-ci.org/browse/INFRA-2794)

File: core/src/main/java/hudson/model/ListView.java
Comment (rsandell) [inline core/src/main/java/hudson/model/ListView.java:135]: Does this imply that all plugins must make the same change?

Comment (timja) [reply core/src/main/java/hudson/model/ListView.java:135]: If the class is extended and the parent read resolve needs to be called

Comment (daniel-beck) [reply core/src/main/java/hudson/model/ListView.java:135]: https://github.com/jenkinsci/jep/tree/master/jep/228#inherited-private-serialization-methods explains this

Comment (jglick) [reply core/src/main/java/hudson/model/ListView.java:135]: I attempted to search for extant plugins using an incorrect `private` override, but I might have missed some of course.

**Pull request merged**
